File: 0061-rotate-list/0061-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(!head || !head->next || k == 0) return head; 
        vector<ListNode*> nums;//to store the nodes of the linked list
        vector<ListNode*> inter;//to reverse the linked list 
        int count = 0; // to find the length of the linked list
        ListNode* temp = head;
        while(temp){
            nums.push_back(temp);
            temp = temp->next; 
        }
        count = nums.size(); 
        k = k % count;// to accomodate for k values that are greater the size of the linked list
        if(k == 0) return head; 
       

        ListNode* Temp = nums[count - k];//mapping the start of the reverse
        nums[count - k - 1]->next = nullptr; 
        for(int i = count - k; i < count; i++){//to find the node before the target, as when we move the target to the first index we need this node to make this node's next to null
            inter.push_back(nums[i]);//adding the last k elements to the vector first 
        } 

        for(int i = 0; i < count - k; i++){
            inter.push_back(nums[i]);// then adding the first elements...  
        }

        inter[k - 1]->next = head;// connecting the last element of inter to the first element of the linked list  

        return Temp; 
        
        

        

        
    }
};

File: 0061-rotate-list/0061-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        #use two pointer, this seems like remove nth node from the end of the list 
        #move the second  ptr k+1 times then move the second pointer until first points outside 
        dummy = ListNode(0, head)
        slow = fast = dummy 
        
        length = 0
        curr = head
        while curr:
            curr = curr.next
            length += 1
        if not head: 
            return None 
        k = k % length
        for _ in range(k+1): 
            fast = fast.next
        
        while fast and fast.next: 
            slow = slow.next
            fast = fast.next 
        
        slow = slow.next

        fast.next = head
        head = slow.next 
        slow.next = None 

        return head